react-app/src/components/SaveOutputForm.js

```diff
@@ -0,0 +1,213 @@
+import React, { useState } from 'react';
+import { Alert, Button, Col, Form, Row } from 'react-bootstrap';
+import fhirpath from 'fhirpath';
+
+function SaveOutputForm(props) {
+  const [outputPath, setOutputPath] = useState('No Folder Selected');
+  const [selectAll, setSelectAll] = useState(true);
+  const [saveLogs, setSaveLogs] = useState(false);
+
+  const [showSavedAlert, setShowSavedAlert] = useState(false);
+  const [showErrorAlert, setShowErrorAlert] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [showNoFilesAlert, setShowNoFilesAlert] = useState(false);
+
+  function onSetOutputPath() {
+    window.api.getOutputPath().then((savePath) => {
+      if (!savePath.canceled) {
+        setOutputPath(savePath.filePaths[0]);
+      }
+    });
+  }
+
+  function clearOutputPath() {
+    setOutputPath('No Folder Selected');
+  }
+
+  function toggleSaveLogs() {
+    setSaveLogs(!saveLogs);
+  }
+
+  function getLabel(bundle, id) {
+    // attempt to get MRN and name with fhirpath
+    const patient = fhirpath.evaluate(bundle, "Bundle.descendants().resource.where(resourceType='Patient')")[0];
+    if (patient) {
+      const mrn = patient.id;
+      const name = patient.name[0].text;
+
+      const isMasked = fhirpath.evaluate(patient, 'Patient.identifier.extension.valueCode')[0] === 'masked';
+
+      // if both MRN and name -- return string w / both
+      if (typeof mrn === 'string' && mrn.length > 0 && typeof name === 'string' && name.length > 0 && !isMasked) {
+        const label = mrn.concat(': ').concat(name);
+        return label;
+      }
+      // if either MRN or name -- return string / the available one
+      if (typeof name === 'string' && name.length > 0) {
+        return name;
+      }
+      // if neither MRN nor name -- "Patient " + patient_resource_id
+      if (typeof mrn === 'string' && mrn.length > 0) {
+        return mrn;
+      }
+    }
+
+    // if no patient resource ID -- return "Patient " + props.id
+    let label = 'Patient';
+    label = label.concat(' ').concat(id.toString());
+    return label;
+  }
+
+  let defaultWhichFiles = {};
+  props.extractedData.forEach((bundle, i) => {
+    const label = getLabel(bundle, i);
+    defaultWhichFiles = {
+      ...defaultWhichFiles,
+      [label]: true,
+    };
+  });
+  const [whichFiles, setWhichFiles] = useState({ ...defaultWhichFiles });
```

Comment:
Using setWhichFiles() triggers a re-render. Using it in the main SaveOutputForm body leads to an infinite loop of re-renders, since each re-render causes the function to be called again. Additionally, the forEach loop that populates defaultWhichFiles must occur after the declaration of getLabel, since getLabel's return value is used to create the keys.
As a result, I was only able to move this section of code (everything from the defaultWhichFiles declaration to the useState call) to the variable declaration section by declaring the getLabel function outside of and before the SaveOutputForm function. Is this okay, or should I leave the code how it was?

Reply:
I want to think about this one a little more, but I have a 12-1 meeting before the working session, so I wanted to post the other comments so you have them. But I'm coming back to this!

Reply:
Okay I reread this and I think it makes sense how it is now! This is basically just setting up how the default value for the variable works, so I think I had gotten confused earlier, but I like it!